File: feature-map/src/main/java/home/oleg/placesnearme/feature_map/di/VenuesMapFragmentModule.java
package home.oleg.placesnearme.feature_map.di;

import android.app.Activity;
import android.arch.lifecycle.LifecycleOwner;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;

import com.smedialink.feature_add_favorite.CreateFavoriteViewModel;
import com.smedialink.feature_venue_detail.venue.viewmodel.VenueViewModel;

import java.util.List;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import home.oleg.feature_add_history.CheckInViewModel;
import home.oleg.placesnearme.feature_map.adapter.CheckedItem;
import home.oleg.placesnearme.feature_map.adapter.SectionsAdapter;
import home.oleg.placesnearme.feature_map.drawable_converter.DrawableConverter;
import home.oleg.placesnearme.feature_map.drawable_converter.DrawableConverterImpl;
import home.oleg.placesnearme.feature_map.sections.SectionProvider;
import home.oleg.placesnearme.feature_map.view.VenuesMapFragment;
import home.oleg.placesnearme.feature_map.viewdata.SectionViewData;
import home.oleg.placesnearme.feature_map.viewmodel.UserLocationViewModel;
import home.oleg.placesnearme.feature_map.viewmodel.VenuesViewModel;
import io.reactivex.annotations.NonNull;

@Module
public abstract class VenuesMapFragmentModule {

    @Provides
    @NonNull
    static Activity provideActivity(VenuesMapFragment fragment) {
        return fragment.getActivity();
    }

    @Provides
    @NonNull
    static CheckInViewModel provideCheckInViewModel(
            VenuesMapFragment fragment,
            ViewModelProvider.Factory factory) {
        return ViewModelProviders.of(fragment, factory).get(CheckInViewModel.class);
    }

    @Provides
    @NonNull
    static CreateFavoriteViewModel provideCreateFavoriteViewModel(
            VenuesMapFragment fragment,
            ViewModelProvider.Factory factory) {
        return ViewModelProviders.of(fragment, factory).get(CreateFavoriteViewModel.class);
    }

    @Provides
    @NonNull
    static VenueViewModel provideVenueViewModel(
            VenuesMapFragment fragment,
            ViewModelProvider.Factory factory) {
        return ViewModelProviders.of(fragment, factory).get(VenueViewModel.class);
    }

    @Provides
    @NonNull
    static VenuesViewModel provideMapViewModel(
            VenuesMapFragment fragment,
            ViewModelProvider.Factory factory) {
        return ViewModelProviders.of(fragment, factory).get(VenuesViewModel.class);
    }

    @Provides
    @NonNull
    static UserLocationViewModel provideUserLocationViewModel(
            VenuesMapFragment fragment,
            ViewModelProvider.Factory factory) {
        return ViewModelProviders.of(fragment, factory).get(UserLocationViewModel.class);
    }

    @Provides
    @NonNull
    static LifecycleOwner provideLifeCycleOwner(VenuesMapFragment mapFragment) {
        return mapFragment;
    }

    @Binds
    @NonNull
    abstract DrawableConverter provideConverter(DrawableConverterImpl impl);

    @Provides
    public static SectionsAdapter provideSectionsAdapter(VenuesMapFragment venuesMapFragment) {
        SectionProvider sectionProvider = new SectionProvider();
        List<CheckedItem<SectionViewData>> seCheckedItems = CheckedItem.wrap(sectionProvider.getSections());
        return new SectionsAdapter(seCheckedItems, venuesMapFragment);
    }
}

File: feature-map/src/main/java/home/oleg/placesnearme/feature_map/di/VenuesMapFragmentModule.kt
package home.oleg.placesnearme.feature_map.di

import android.app.Activity
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.multibindings.IntoMap
import home.oleg.placesnearme.coredi.mapkeys.ViewModelKey
import home.oleg.placesnearme.coredomain.interactors.GetRecommendedVenues
import home.oleg.placesnearme.coredomain.repositories.UserLatLngRepository
import home.oleg.placesnearme.corettools.error_handler.ErrorHandler
import home.oleg.placesnearme.feature_add_favorite.presentation.CreateFavoriteViewModelDelegate
import home.oleg.placesnearme.feature_add_history.presentation.viewmodel.CheckInViewModelDelegate
import home.oleg.placesnearme.feature_map.drawable_converter.DrawableConverter
import home.oleg.placesnearme.feature_map.drawable_converter.DrawableConverterImpl
import home.oleg.placesnearme.feature_map.sections.SectionProvider
import home.oleg.placesnearme.feature_map.presentation.MapFragment
import home.oleg.placesnearme.feature_map.presentation.adapter.CheckedItem
import home.oleg.placesnearme.feature_map.presentation.adapter.SectionsAdapter
import home.oleg.placesnearme.feature_map.presentation.MapViewModel
import home.oleg.placesnearme.feature_venue_detail.domain.GetDetailedVenue
import home.oleg.placesnearme.feature_venue_detail.presentation.VenueViewModel

@Module
abstract class VenuesMapFragmentModule {

    @Binds
    internal abstract fun provideConverter(impl: DrawableConverterImpl): DrawableConverter

    @Module
    companion object {

        @JvmStatic
        @Provides
        internal fun provideActivity(fragment: MapFragment): Activity {
            return fragment.activity!!
        }

        @JvmStatic
        @IntoMap
        @ViewModelKey(VenueViewModel::class)
        @Provides
        internal fun provideVenueViewModel_(errorHanlder: ErrorHandler,
                                            createFavoriteViewModelDelegate: CreateFavoriteViewModelDelegate,
                                            checkInViewModelDelegate: CheckInViewModelDelegate,
                                            getDetailedVenue: GetDetailedVenue): ViewModel {
            return VenueViewModel(createFavoriteViewModelDelegate,
                    checkInViewModelDelegate,
                    errorHanlder,
                    getDetailedVenue)
        }

        @JvmStatic
        @IntoMap
        @ViewModelKey(MapViewModel::class)
        @Provides
        internal fun provideMapViewModel_(errorHandler: ErrorHandler,
                                          interactor: GetRecommendedVenues,
                                          userLocationRepository: UserLatLngRepository): ViewModel {
            return MapViewModel(errorHandler, userLocationRepository, interactor)
        }

        @JvmStatic
        @Provides
        internal fun provideVenueViewModel(
                fragment: MapFragment,
                factory: ViewModelProvider.Factory): VenueViewModel {
            return ViewModelProviders.of(fragment, factory).get(VenueViewModel::class.java)
        }

        @JvmStatic
        @Provides
        internal fun provideMapViewModel(
                fragment: MapFragment,
                factory: ViewModelProvider.Factory): MapViewModel {
            return ViewModelProviders.of(fragment, factory).get(MapViewModel::class.java)
        }

        @JvmStatic
        @Provides
        internal fun provideSectionsAdapter(venuesMapFragment: MapFragment): SectionsAdapter {
            val sectionProvider = SectionProvider()
            val seCheckedItems = CheckedItem.wrap(sectionProvider.sections)
            return SectionsAdapter(seCheckedItems, venuesMapFragment)
        }
    }
}
